nexus/server.go

```diff
@@ -0,0 +1,471 @@
+	return &OperationResponseSync{
+		Header: header,
+		Body:   bytes.NewReader(b),
+	}, nil
+}
+
+func (r *OperationResponseSync) applyToStartResponse(writer http.ResponseWriter, handler *httpHandler) {
+	header := writer.Header()
+	for k, v := range r.Header {
+		header[k] = v
+	}
+	if closer, ok := r.Body.(io.Closer); ok {
+		defer closer.Close()
+	}
+	if _, err := io.Copy(writer, r.Body); err != nil {
+		handler.logger.Error("failed to write response body", "error", err)
+	}
+}
+
+func (r *OperationResponseSync) applyToGetResultResponse(writer http.ResponseWriter, handler *httpHandler) {
+	writer.Header().Set(headerOperationState, string(OperationStateSucceeded))
+	r.applyToStartResponse(writer, handler)
+}
+
+// Indicates that an operation has been accepted and will complete asynchronously.
+type OperationResponseAsync struct {
+	OperationID string
+}
+
+func (r *OperationResponseAsync) applyToStartResponse(writer http.ResponseWriter, handler *httpHandler) {
+	info := OperationInfo{
+		ID:    r.OperationID,
+		State: OperationStateRunning,
+	}
+	bytes, err := handler.marshaler(info)
+	if err != nil {
+		handler.logger.Error("failed to serialize operation info", "error", err)
+		writer.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	writer.Header().Set(headerContentType, contentTypeJSON)
+	writer.WriteHeader(http.StatusCreated)
+
+	if _, err := writer.Write(bytes); err != nil {
+		handler.logger.Error("failed to write response body", "error", err)
+	}
+}
+
+func (r *OperationResponseAsync) applyToGetResultResponse(writer http.ResponseWriter, handler *httpHandler) {
+	writer.Header().Set(headerOperationState, string(OperationStateRunning))
+	writer.WriteHeader(statusOperationRunning)
+}
+
+// A Handler must implement all of the Nexus service endpoints as defined in the [Nexus HTTP API].
+//
+// Handler implementations must embed the [UnimplementedHandler].
+//
+// All Handler methods can return a [HandlerError] to fail requests with a custom status code and structured [Failure].
+//
+// [Nexus HTTP API]: https://github.com/nexus-rpc/api
+type Handler interface {
+	// StartOperation handles requests for starting an operation. Return [OperationResponseSync] to respond successfully
+	// - inline, or [OperationResponseAsync] to indicate that an asynchronous operation was started.
+	// Return an [UnsuccessfulOperationError] to indicate that an operation completed as failed or canceled.
+	StartOperation(context.Context, *StartOperationRequest) (OperationResponse, error)
+	// GetOperationResult handles requests to get the result of an asynchronous operation. Return
+	// [OperationResponseSync] to respond successfully - inline, or [OperationResponseAsync] to indicate that an
+	// asynchronous operation is still running.
+	// Return an [UnsuccessfulOperationError] to indicate that an operation completed as failed or canceled.
+	GetOperationResult(context.Context, *GetOperationResultRequest) (OperationResponse, error)
```

Comment:
@cretz 
I know what you're going to say, you want this to return an error on operation not yet completed.

I tried that but I prefer the consistency with `StartOperation` here.
Note that I also treat DeadlineExceeded with `Wait=true` as `OperationResponseAsync` to simplify handler implementation.

Reply:
I'll post that version in the next commit and we can decide which one to keep.